// @angularclass/webpack-toolkit/dist/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var core_1 = require('@angular/core');
var compiler_1 = require('@angular/compiler');
exports.AC_WEBPACK_ASYNC_MAP = {};
exports.ANGULARCLASS_WEBPACK_RUNTIME_PROVIDERS = [
    { provide: exports.AC_WEBPACK_ASYNC_MAP, useValue: {} },
    {
        provide: core_1.ComponentResolver,
        useFactory: function (resolver, asyncMap) {
            return new WebpackComponentResolver(resolver, asyncMap);
        },
        deps: [compiler_1.RuntimeCompiler, exports.AC_WEBPACK_ASYNC_MAP]
    }
];
function provideWebpack(config) {
    return exports.ANGULARCLASS_WEBPACK_RUNTIME_PROVIDERS.concat([
        { provide: exports.AC_WEBPACK_ASYNC_MAP, useValue: config },
    ]);
}
exports.provideWebpack = provideWebpack;
var WebpackComponentResolver = (function () {
    function WebpackComponentResolver(_resolver, _asyncComponents) {
        this._resolver = _resolver;
        this._asyncComponents = _asyncComponents;
    }
    WebpackComponentResolver.prototype.resolveComponent = function (componentType) {
        if (typeof componentType === 'string' && this._asyncComponents[componentType]) {
            return this._fetchComponent(componentType);
        }
        return this._resolver.resolveComponent(componentType);
    };
    WebpackComponentResolver.prototype.prefetchComponent = function (componentType) {
        return this._fetchComponent(componentType);
    };
    WebpackComponentResolver.prototype.clearCache = function () { };
    WebpackComponentResolver.prototype._fetchComponent = function (componentType) {
        var _this = this;
        return this._asyncComponents[componentType](componentType)
            .then(function (cmp) {
            return _this._resolver.resolveComponent(cmp);
        });
    };
    WebpackComponentResolver = __decorate([
        core_1.Injectable(),
        __param(1, core_1.Inject(exports.AC_WEBPACK_ASYNC_MAP)), 
        __metadata('design:paramtypes', [core_1.ComponentResolver, Object])
    ], WebpackComponentResolver);
    return WebpackComponentResolver;
}());
exports.WebpackComponentResolver = WebpackComponentResolver;
